---
title: "Week 3: Genetic Diversity"
author: "Helene Wagner"
date: "`r Sys.Date()`"
show_toc: true
output:
  knitr:::html_vignette:
    toc: yes
    fig_width: 4 
    fig_height: 3.5
vignette: >
  %\VignetteIndexEntry{Week 3: Genetic Diversity}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## 1. Overview of Worked Example

### a) Goals 

This worked example shows how to:

- Check markers and populations (polymorphism, HWE, linkage, null alleles). 
- Add sampling locations and hierarchical structure to genind object
- Assess genetic diversity and inbreeding.
- Calculate genetic distances among individuals and populations.
- Analyze genetic structure with F statistics (and alternatives) and AMOVA.

### b) Data set 

This is the same data set as used in Weeks 1 & 2.

Microsatellite data for 181 individuals of Colombia spotted frogs (Rana luteiventris) from 12 populations. Site-level spatial coordinates and attributes. The data are a subsample of the full data set analyzed in Funk et al. (2005) and Murphy et al. (2010). Please see the separate introduction to the data set. 

- **ralu.loci**: Data frame with populations and genetic data (181 rows x 9 columns). Included in package 'LandGenCourse'. To load it, type: data(ralu.loci)
- **ralu.site**: Spatial points data frame with spatial coordinates and site variables Included in package GeNetIt'. To load it, type: data(ralu.site)

### c) Required R libraries

All required packages should have been installed already when you installed 'LandGenCourse'.

```{r message=FALSE, warning=TRUE}
require(adegenet)
require(gstudio)
require(LandGenCourse)
require(pegas)       
require(sp)
require(hierfstat)  
require(PopGenReport)
require(GeNetIt)
require(dplyr)
require(poppr)       
require(mmod)
```

## 2. Basic checking of markers and populations

Before we do landscape genetic analysis, we need to perform a basic population genetic analysis of the genetic data, in order to better understand the nature and quality of the data and to check for underlying assumptions of population genetic models and corresponding methods. 

### a) Re-create genind object 

Adapted from Week 1 tutorial: 

```{r}
data(ralu.loci, package="LandGenCourse")
Frogs <- data.frame(FrogID = paste(substr(ralu.loci$Pop, 1, 3), 
                                   row.names(ralu.loci), sep="."), ralu.loci)
Frogs.genind <- adegenet::df2genind(X=Frogs[,c(4:11)], sep=":", ncode=NULL, 
                          ind.names= Frogs$FrogID, loc.names=NULL, 
                          pop=Frogs$Pop, NA.char="NA", ploidy=2, 
                          type="codom", strata=NULL, hierarchy=NULL)
Frogs.genind
```

### b) Check that markers are polymorphic
The genetic resolution depends on the number of markers and their polymorphism. The table above and the summary function for genind objects together provide this information. Here we have:

- 8 loci with 3 - 9 alleles (39 in total)
- Expected heterozygosity varies between 0.14 (locus C) and 0.78 (locus E)
- There's a reasonable level of missing values (10.6%) 

```{r}
summary(Frogs.genind)
```

### c) Check for deviations from Hardy-Weinberg equilibrium (HWE)

See also: http://dyerlab.github.io/applied_population_genetics/hardy-weinberg-equilibrium.html

For a very large population (no drift) with random mating and non-overlapping generations (plus a few more assumptions about the mating system), and in the absence of mutation, migration (gene flow) and selection, we can predict offspring genotype frequencies from allele frequencies of the parent generation (Hardy-Weinberg equilibrium). 
In general, we don't expect all of these assumptions to be met (e.g., if we want to study gene flow or selection, we kind of expect that these processes are present). Note: plants often show higher levels of departure from HWE than animals. 

Here are p-values for two alternative tests of deviation from HWE for each locus. Columns:

- **chi^2**: value of the classical chi-squared test statistic
- **df**: degrees of freedom of the chi-squared test
- **Pr(chi^2 >)**: p-value of the chi-squared test ('>' indicates that the alternative is 'greater', which is always the case for a chi-squared test)
- **Pr.exact**: p-value from an exact test based on Monte Carlo permutation of alleles (for diploids only). The default is B = 1000 permutations (set B = 0 to skip this test). 

```{r}
round(pegas::hw.test(Frogs.genind, B = 1000),3)
```

Both tests suggest that most loci are out of HWE globally (across all 181 individuals). Next, we check for HWE of each locus in each population.

Note on code: The curly brackets '{ }' below are used to keep the output from multiple lines together in the html file.

```{r}
# Chi-squared test: p-value
HWE.test <- data.frame(sapply(seppop(Frogs.genind), 
                              function(ls) pegas::hw.test(ls, B=0)[,3]))
HWE.test.chisq <- t(data.matrix(HWE.test))
{cat("Chi-squared test (p-values):", "\n")
round(HWE.test.chisq,3)}
```

```{r}
# Monte Carlo: p-value
HWE.test <- data.frame(sapply(seppop(Frogs.genind), 
                              function(ls) pegas::hw.test(ls, B=1000)[,4]))
HWE.test.MC <- t(data.matrix(HWE.test))
{cat("MC permuation test (p-values):", "\n")
round(HWE.test.MC,3)}
```
To summarize, let's calculate, for each locus, the proportion of populations where it was out of HWE. Here we'll use the conservative cut-off of alpha = 0.05 for each test. There are various ways of modifying this, including a simple Bonferroni correction, where we divide alpha by the number of tests, which you can activate here by removing the # in front of the line.

```{r}
alpha=0.05
#alpha=0.05/(nrow(HWE.test) * ncol(HWE.test))   # Optional Bonferroni correction
Prop.loci.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 2, mean), 
           MC=apply(HWE.test.MC<alpha, 2, mean))
Prop.loci.out.of.HWE
```

And similarly, for each population, the proportion of loci that were out of HWE:

```{r}
Prop.pops.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 1, mean), 
           MC=apply(HWE.test.MC<alpha, 1, mean))
Prop.pops.out.of.HWE
```
The results suggest that:

- While most loci are out of HWE globally, this is largely explained by subdivision (variation in allele frequencies among local populations indicating limited gene flow). 
- No locus is consistently out of HWE across populations (loci probably not affected by selection).
- No population is consistently out of HWE across loci (probably no recent major bottlenecks/ founder effects).

### d) Check for linkage disequilibrium (LD)
See also: https://grunwaldlab.github.io/Population_Genetics_in_R/Linkage_disequilibrium.html

For microsatellite markers, we typically don't know where on the genome they are located. The closer together two markers are on a chromosome, the more likely they are inherited together, which means that they don't really provide independent information. Testing for linkage disequilibrium assesses this, for each pair of loci, by checking whether alleles of two loci are statistically associated.

This step is especially important when developing a new set of markers. You may want to drop (the less informative) one marker of any pair of linked loci. 

Here, we start with performing an overall test of linkage disequilibrium (the null hypothesis is that there is no linkage among the set of markers). Two indices are calculated and tested: an index of association (Ia; Brown et al. 1980) and a measure of correlation (rbarD; Agapow and Burt 2001), which is less biased (see URL above). 

Overall, there is statistically significant association among the markers (p-value: prD = 0.005; also left figure). Recall that the power of a statistical increases with sample size, and here we have n = 181, hence even a small effect may be statistically significant. Hence we look at effect size, i.e., the actual strength of the pairwise associations (right figure). 

```{r, fig.show='hold'}
poppr::ia(Frogs.genind, sample=199)
LD.pair <- poppr::pair.ia(Frogs.genind)
```

The strongest correlation is around 0.2, for markers E and H. 

Effect size: If rbarD can be interpreted similarly to a linear correlation coefficient r, that would mean that less than 5% of the variation in one marker is shared with the other marker (recall from stats: the amount of variance explained in regression, Rsquared, is the square of the linear correlation coefficient). This is probably not large enough to worry about.  

### e) Check for null alleles (PopGenReport)

NEED SOME TEXT HERE

The function 'null.all' returns a list with two components ('homozygotes' and 'null.allele.freq'), and each of these is again a list. See '?null.all' for details and choice of method.

List 'homozygotes':
  
- **homozygotes$observed**: observed number of homozygotes for each allele at each locus
- **homozygotes$bootstrap**: distribution of the expected number of homozygotes
- **homozygotes$probability.obs**: probability of observing the number of homozygotes

```{r}
# Null alleles: depends on method! See help file.
Null.alleles <- PopGenReport::null.all(Frogs.genind)
Null.alleles$homozygotes$probability.obs
```

List 'null.allele.freq': 
  
- **null.allele.freq$summary1**: null allele frequency estimates based upon the forumulas of Chakraborty et al. (1994)
- **null.allele.freq$summary2**: null allele frequency estimates based upon the forumulas of Brookfield (1996)

From the help file: "Brookfield (1996) provides a brief discussion on which estimator should be used. In summary, it was recommended that Chakraborty et al. (1994)'s method (e.g. summary1) be used if there are individuals with no bands at a locus seen, but they are discounted as possible artefacts. If all individuals have one or more bands at a locus then Brookfield (1996)'s method (e.g. summary2) should be used." In this case, we have many individuals with missing values for both alleles, hence better use summary1.

Each summary table contains a summary with observed, median, 2.5th percentile and 97.5the percentile. The percentiles form a 95% confidence interval. From the help file: "If the 95% confidence interval includes zero, it indicates that the frequency of null alleles at a locus does not significantly differ from zero."

Both methods suggest that there may be null alleles in most (summary1) or all loci (summary2).

WHAT TO DO WITH THIS INFORMATION??
```{r}
{cat(" summary1 (Chakraborty et al. 1994):", "\n")
round(Null.alleles$null.allele.freq$summary1,2)  
cat("\n", "summary2 (Brookfield et al. 1996):", "\n")
round(Null.alleles$null.allele.freq$summary2,2)}   
```

## 3. Add sampling locations and hierarchical structure to genind object

The 'genind' object can store non-genetic data in a slot called 'other'. Here we'll add spatial coordinates and site data, including the variables that define the hierarchical sampling design, where individual frogs were sampled at ponds (local populations), which are nested within basins, which are nested within drainages.

- **other$latlong**: coordinates in decimal lat/lon format
- **other$xy**: coordinates in UTM format
- **other**: may add any attribute data. Here we'll add the site data from ralu.site.

### a) Add spatial coordinates and site variables to 'other' slot

We import the UTM coordinates from 'ralu.site@coords'. We use the function 'spTransform' to conver them to latlong data (x = Longitude, y = Latitude) to facilitate plotting results on Google maps later in the course. We use 'get_proj4' (see Week 2) to define the projection.

```{r}
data(ralu.site, package="GeNetIt")
coords.longlat <- sp::spTransform(ralu.site, tmaptools::get_proj4("longlat"))@coords
dimnames(coords.longlat)[[2]] <- c("Longitude", "Latitude")
```

The coordinates and site data are at the site level (and include more than our 12 sites), whereas the genetic data are at the individual level (181 frogs). Here we use the function 'left_join' from the 'dplyr' package to extract the coordinates and site data for each frog into a new dataframe 'tmp'.

The function 'left_join' takes a table x (here: columns 1:2 of 'ralu.loci', which has the same order as the data in our genind object that was created from it) and a table y (here, we bind the columns from the tables with the latlong coordinates, the UTM coordinates, and the site data from ralu.site into a single dataframe, using function 'cbind'). The arguent 'by' specifies which column in x and in y should be matched). The function 'left_join' then finds the rows in table y that match the rows in table x. Hence the resulting table 'tmp' should have 181 rows, one for each frog.

```{r}
tmp <- dplyr::left_join(x=ralu.loci[,1:2], 
                 y=cbind(coords.longlat, ralu.site@coords, ralu.site@data), 
                 by = c("SiteName" = "SiteName"))
{cat("Number of rows and columns in joined table:", "\n")
dim(tmp)
cat("Variable names:", "\n")
names(tmp)}
```

Then we write the latlong coordinates into 'latlong', the UTM coordinates into 'xy', and the site data into 'site' (we can give this one any name). Disregard the warning: 'SiteName' is a factor in one dataset and character in the other, R is taking care of this. (We'll look at the result after the next step).

```{r}
Frogs.genind@other$latlong <- tmp[,3:4]
Frogs.genind@other$xy <- tmp[,5:6]
Frogs.genind@other$site <- tmp[,-c(3:6)]
```

### b) Fill 'strata' and 'hierarchy' slots

Defining the sampling hierarchy is required e.g. for testing for significant differentiation at each level (AMOVA). We need to define two slots:

- **strata**: this is a table with all variables needed to define the hierarchical levels
- **hierarchy**: this is a formula representation of how the levels are nested. 

Here, ponds ('Pop') are nested within basins ('Basin'), which are nested within drainage systems ('Drainage'). The corresponding hierarchy formula is: ~ Drainage/Basin/Pop

The tilde symbol '~' generally indicates that something (here the genetic data) is modeled as a function of the terms to the right of the symbol. The slash symbol '/' indicates how the hierarchical levels are nested, starting from the top level. 

```{r}
# Define strata
adegenet::strata(Frogs.genind) <- with(Frogs.genind@other$site, data.frame(Drainage, Basin, SiteName, Pop))
# Define hierarchy
adegenet::hier(Frogs.genind) <- ~ Drainage/Basin/Pop
```

Check that the slots 'strata', 'hierarchy' and 'other' are correctly filled:

```{r}
Frogs.genind
```

### c) Aggregate genetic data at population level (allele frequencies)

For some analyses, we will need to aggregate data from the individual to the population level, e.g. as a table of allele frequencies per population.

Here we convert the 'genind' object to a 'genpop' object (NOT the same as a 'genepop' object!). This is defined in the package 'adegenet' to hold population-level genetic data. The function 'genind2genpop' obviously converts from 'genind' to 'genpop'.

```{r}
Frogs.genpop <- adegenet::genind2genpop(Frogs.genind)
```

The function 'makefreq' extracts the table with allele frequencies from the 'genpop' object. We'll plot just a few lines and alleles. 

```{r}
Freq <- adegenet::makefreq(Frogs.genpop)
round(Freq[1:6,1:10], 2)
```

The allele frequencies of all alleles from the same locus (e.g., A.1, A.2 and A.3) should sum to 1 for each population. With eight loci, the row sums should thus add to 8. Here we apply the function 'sum' to each row (MARGIN = 1, would be 2 for columns) of the table of allele frequencies 'Freq'.

```{r}
apply(Freq, MARGIN = 1, FUN = sum)    # Just checking
```

## 4. Assess genetic diversity and inbreeding (adegenet)

These measures are typically quantified per population. 

### a) Rarefied allelic richness

Nominal sample size (number of frogs sampled) and valid sample size (e.g., for each locus, the number of frogs with non-missing genetic data) vary between sites. Hence we should not compare allelic richness directly. Instead, we'll use rarefied allelic richness (Ar). 

By default, the function 'allel.rich' finds the lowest valid sample size across all populations and loci, and multiplies it by the ploidy level. The number is stored as 'Richness$alleles.sampled' (here: 3 individuals * 2 alleles = 6 alleles). Alternatively, this number can be set with the 'min.alleles' argument.  

Populations with more alleles are resampled to determine the average allelic richness among the minimum number of allels. Here. this means that 6 alleles are sampled from each population, allelic richness is calculated, and the process is repeated many times to determine the average).

```{r}
# rarefied allelic richness
Richness <- PopGenReport::allel.rich(Frogs.genind, min.alleles = NULL)
Richness
```  
Note: 'pop.sizes' here refers to the number of (non-missing) alleles, not individuals.

Let's plot the results. The barchart shows that there is considerable variation in genetic diversity among ponds. A scatterplot against sample size (here: for each population, the average number of valid alleles across loci) suggests that the variation is not related to sample size. The regression line (red) is almost horizontal.

```{r, fig.show='hold'}
par(mar=c(5.5, 4.5,1,1))
barplot(Richness$mean.richness, las=3, ylab="Rarefied allelic richness (Ar)")
plot(colMeans(Richness$pop.sizes), Richness$mean.richness,
     xlab="Valid sample size", 
     ylab="Rarefied allelic richness (Ar)")
abline(lm(Richness$mean.richness ~ colMeans(Richness$pop.sizes)), col="red")

```

### b) Observed and expected heterozygosity

Note: Writing the 'genind' summary into an object 'Sum' allows accessing its attributes by name.

```{r}
  # Summarize data
  Sum <- summary(Frogs.genind)
  names(Sum)
```
  
Expected heterozygosity (Hexp) is a measure of genetic diversity among individuals, and observed heterozygosity (Hobs) within individuals. Here are the global values (pooled across all populations):

```{r, fig.show='hold', fig.height=3, fig.width=4}
  par(mar=c(3, 4.5,1,1))
  # Observed and expected heterozygosity by locus
  barplot(Sum$Hexp, ylim=c(0,1), ylab="Expected heterozygosity")
  barplot(Sum$Hobs, ylim=c(0,1), ylab="Observed heterozygosity")
``` 

By locus and population:

```{r}
  # Observed and expected heterozygosity by locus and population
  Hobs <- t(sapply(seppop(Frogs.genind), function(ls) summary(ls)$Hobs))
  Hexp <- t(sapply(seppop(Frogs.genind), function(ls) summary(ls)$Hexp))
  {cat("Expected heterozygosity (Hexp):", "\n")
  round(Hexp, 2)
  cat("\n", "Observed heterozygosity (Hobs):", "\n")
  round(Hobs, 2)}
  # 1 - Hobs/Hexp
``` 
Locus F shows variation only in two populations (i.e., Hexp = 0 in 10 populations). 

Let's plot the average across all loci for each population:

```{r, fig.show='hold'} 
  par(mar=c(5.5, 4.5, 1, 1))
  # Observed and expected heterozygosity by population
  Hobs.pop <- apply(Hobs, 1, mean)
  Hexp.pop <- apply(Hexp, 1, mean) 
  barplot(Hexp.pop, ylim=c(0,1), las=3, ylab="Expected heterozygosity")
  barplot(Hobs.pop, ylim=c(0,1), las=3, ylab="Observed heterozygosity")
```  

### c) Inbreeding estimation (adegenet)
See also: http://adegenet.r-forge.r-project.org/files/tutorial-basics.pdf

ADD SOME TEXT

Maximum likelihood estimate of inbreeding coefficient F:

```{r} 
Fest <- lapply(seppop(Frogs.genind), 
               function(ls) adegenet::inbreeding(ls, N=200, res.type="estimate"))
```

Average sample estimates of F: 

```{r}
temp <- lapply(seppop(Frogs.genind), 
               function(ls) adegenet::inbreeding(ls, N=200, res.type="sample"))
Fbar <- lapply(temp, function (ls) sapply(ls, mean))
```
The two values can be quite different:

```{r, fig.show='hold'}
par(mar=c(5.5, 4.5, 1, 1))
boxplot(Fest, las=3, ylim=c(0,1), xlab="", ylab="Inbreeding coefficient (Fest)")
boxplot(Fbar, las=3, ylim=c(0,1), xlab="", ylab="Inbreeding coefficient (Fbar)")
Mean.inbreeding.per.pop <- sapply(Fbar, mean)
```  


## 5. Calculate individual- and population-level genetic distances

The degree of genetic differentiation between two populations reflects the amount of gene flow between them. Different measures of genetic distance can be used to quantify the differentiation (i.e., how different the allele frequencies are), either between pairs of individuals or pairs of populations. 

This results in a distance matrix of size n x n (here, 181 x 181 for individual-level and 12 x 12 for population-level analysis). An alternative representation is an object of class 'dist', which contains only the (n*(n - 1)/2) values of the lower triangle of the distance matrix. 

### a) Genetic distance among individuals (PopGenReport, adegenet)

```{r}
# Individual-level genetic distances
GD.ind.smouse <- PopGenReport::gd.smouse(Frogs.genind, verbose=FALSE)  # GenAlEx
GD.ind.propShared <- adegenet::propShared(Frogs.genind)     
```

### b) Genetic distance among local populations (PopGenReport, adegenet, mmod)

```{r}
# Population-level genetic distances
GD.pop.propShared <- PopGenReport::pairwise.propShared(Frogs.genind)
GD.pop.Nei <- adegenet::dist.genpop(Frogs.genpop, method=1)   
GD.pop.Edwards <- adegenet::dist.genpop(Frogs.genpop, method=2)
GD.pop.Reynolds <- adegenet::dist.genpop(Frogs.genpop, method=3)  # Co-ancestry coef
GD.pop.Rogers <- adegenet::dist.genpop(Frogs.genpop, method=4)  
GD.pop.Provesti <- adegenet::dist.genpop(Frogs.genpop, method=5)

GD.pop.Joost <- mmod::pairwise_D(Frogs.genind, linearized = FALSE)
GD.pop.Hedrick <- mmod::pairwise_Gst_Hedrick(Frogs.genind, linearized = FALSE)
GD.pop.NeiGst <- mmod::pairwise_Gst_Nei(Frogs.genind, linearized = FALSE)
```
The argument 'linearized = TRUE' will linearize each distance measure x by calculating x/(1 - x). This should result in more linear relationships when plotted or correlated against geographic distance.

 
### c) Pair-wise Fst (hierfstat)

Pair-wise Fst values are calculated at the population level.

```{r}
Pairwise.fst <- hierfstat::pairwise.fst(Frogs.genind, pop = NULL, 
                                        res.type = c("dist"))
```

### d) Export distance matrices

We'll need the genetic distance matrices again in later labs, so let's export them. Adding the package name to each matrix name helps keeping track of methods used.

Technically speaking, the two proportion of shared alleles matrices contain similarity values, not distance values. They are proportions between 0 - 1, and we can convert them to distances by subtracting them from 1. 

Note: additional distance measures are available in the package 'gstudio'. Specifically, we will use conditional genetic distance later in the course.

```{r}
GD.ind <- list(smouse.PopGenReport = GD.ind.smouse,
               propShared.adegent = 1 - GD.ind.propShared)

GD.pop <- list(pairwiseFst.hierfstat = Pairwise.fst,
               propShared.PopGenReport = 1 - GD.pop.propShared,
               Nei.adegenet = GD.pop.Nei,
               Edwards.adegenet = GD.pop.Edwards,
               Reynolds.adegent = GD.pop.Reynolds,
               Rogers.adegenet = GD.pop.Rogers,
               Provesti.adegent = GD.pop.Provesti,
               Joost.mmod = GD.pop.Joost,
               Hedrick.mmod = GD.pop.Hedrick,
               Nei.mmod = GD.pop.NeiGst)
```

round(cor(sapply(GD.ind, function(ls) as.vector(ls))),2)
round(cor(sapply(GD.pop, function(ls) as.vector(ls))),2)

Use 'save' to save an R object to your file system, and 'load' to read it in again. Note: the default setting is that 'save' will overwrite existing files with the same name.

```{r}
save(GD.ind, file = "./data/GD.ind.RData")
load("./data/GD.ind.RData")

save(GD.pop, file = "./data/GD.pop.RData")
load("./data/GD.pop.RData")
```

### e) Test for isolation by distance (adegenet)

Let's do a quick-and-dirty check for isolation by distance (IBD). For this, we need to extract the UTM coordinates per population. Here we use functions from the package 'dplyr' to group coordinates by population, then take their mean (they should be identical anyways).

```{r}
coords.pop <- Frogs.genind@other$xy %>%
  group_by(Frogs.genind@other$site$Pop) %>%
  summarize(x = mean(coords.x1, na.rm = TRUE),
            y = mean(coords.x2, na.rm = TRUE))
```

Now we can calculate geographic distance with the 'dist' function.

```{r}
Dgeo <- dist(coords.pop[,-1])

```

We'll cycle through all population-level distance matrices. For each one, we carry out a Mantel test, and write the value of the Mantel r statistic, and its p-value, into an object 'Result'.

```{r}
Result <- matrix(NA, length(GD.pop), 2, 
                 dimnames=list(names(GD.pop), c("Mantel.r", "p-value")))
for(i in 1:length(GD.pop))
{
  IBD <- mantel.randtest(as.dist(GD.pop[[i]]),Dgeo)
  Result[i,] <- c(IBD$obs, IBD$pvalue)
}
Result
```      

The Mantel r statistics range between 0.64 and 0.74, which indicates quite strong spatial structure. All tests were statistically significant at alpha = 0.05, i.e., the null hypothesis of no spatial structure could be rejected no matter which distance measure was used. 

Note that we did not check, or linearize, the shape of the relationships - we'll do a more thorough job in Week 6.

## 6. Analyze genetic structure with F statistics and AMOVA

Beyond distance effects, we may expect mountain ridges to act as barriers for gene flow. The highest ridges typically occur between drainage systems, with minor ridges between basins with a drainage system. We can use Analysis of Molecular Variance to test for such hierarchical structure. This is done in a spatially implicit way, without considering actual spatial locations or geographic distances.

### a) F statistics and analogue measures of genetic differentiation

Before considering the hierarchical sampling design, let's quantify the overall degree of genetic differentiation among populations using F statistics, specifically Fst. 

```{r}

# Overall F statistics (hierfstat)
hierfstat::fstat(Frogs.genind, pop = NULL, fstonly = FALSE)
``` 
How to read this result: COMPLETE THIS!

- pop & Total (top left): Fst - this is a measure of the differentiation among all populations
- Ind & Total (top right): Fit -
- Ind & pop (bottom right): Fis - 

```{r}
# F statistics by locus (pegas)
Frogs.pegas <- pegas::genind2loci(Frogs.genind)
pegas::Fst(Frogs.pegas)

# NOTE: VERY DIFFERENT RESULTS WHEN USING MEAN OF PEGAS, SHOULD BE THE SAME)
# NEED TO CHECK ORDER?
apply(pegas::Fst(Frogs.pegas), 2, mean)
``` 

See also: https://cran.r-project.org/web/packages/mmod/vignettes/mmod-demo.pdf

```{r}
# Fst analogues (mmod)
mmod::diff_stats(Frogs.genind)
```

### b) AMOVA (using 'poppr' interface to 'ade4' and 'pegas' implementations)
See also: https://grunwaldlab.github.io/Population_Genetics_in_R/AMOVA.html

There are multiple implementations of AMOVA in R. The package 'poppr' provides an interface to calculate AMOVA from a 'genind' object using either the implementation in 'ade4' or in 'pegas'. See also: 

Here we'll use the 'ade4' implementation  using the 'poppr.amova' function. Some arguments (see help file for more detail):

- hier: Specify the hierarchy, either by pointing to the 'hierarchy' slot of the 'genind' object, or by making it explicit as 'hier = ~ Drainage/Basin/Pop' (without the quotes).
- within: If TRUE, variation within individuals (must be at least diploid) is assessed.
- dist: optionally, provide an individual-level distance matrix. (Use 'within=FALSE'). The default is to calculate (1 minus) the proportion of shared alleles.
- missing: Options for dealing with missing values. The default is 'loci', which drops any loci with more than 'cutoff' proportion of missing values.
- cutoff: See 'missing'. The default is 0.05. In this example, four of eight loci would be dropped, hence I raised the cutoff to 0.5 to keep all loci. This may affect the result.
- method: either 'ade4' or 'pegas'. See help file ('?poppr.amova').
- nperm: this is only used for the permutation test in 'pegas', not 'ade4' (for ade4, need to run test separately, see code below).

TOTALLY DIFFERENT STATISTICAL TEST RESULTS FOR PEGAS.

```{r}
#amova.result.pegas <- poppr::poppr.amova(Frogs.genind, hier = ~ Drainage/Basin/Pop, 
#            clonecorrect = FALSE, within = FALSE,
#  dist = NULL, squared = FALSE, correction = "quasieuclid", sep = "_",
#  filter = FALSE, threshold = 0, algorithm = "farthest_neighbor",
#  missing = "loci", cutoff = 0.5, quiet = FALSE, method = c(
#  "pegas"), nperm = 200)

amova.result.ade4 <- poppr::poppr.amova(Frogs.genind, hier = Frogs.genind@hierarchy, 
            clonecorrect = FALSE, within = TRUE,
  dist = NULL, squared = FALSE, correction = "quasieuclid", sep = "_",
  filter = FALSE, threshold = 0, algorithm = "farthest_neighbor",
  missing = "loci", cutoff = 0.5, quiet = FALSE, method = c(
  "ade4"), nperm = 0)

amova.result.ade4
```
The effect sizes are under '$componentsofcovariance' in column '%'. Drainage and Basin each explain about 10% of the genetic variation, whereas most variation is within individuals (i.e., unexplained by the hierarchy). Let's test the components with a permutation test.

```{r}
amova.test <- ade4::randtest(amova.result.ade4, nrepet=199) 
amova.test
```

Let's look at the numerical test results:

- There is statistically significant differentiation between drainage systems and between basins, and within samples (p-values < 0.05). 
- The p-values for differentiation between populations, and between individuals (samples), are close to 0.05 and may be too close to call. Note that the exact value may vary from one run of the permutation test to the next. With only 199 replicates, the resolution of the p-value is 1/200 = 0.005. We should redo the test with a larger number of permutations (e.g. 499 or 999). This will take longer!
- Results with nrep=499: p-value < 0.05 for differentiation among populations, >0.05 among individuals.

Hence there is statistically significant differentiation at all levels, except for among frogs sampled at the same pond. The effects are largest for basin and drainage (about 10% each), whereas differentiation among populations within basins is very weak (about 1.5%).

### c) Contribution of each population to Fst

THIS PART IS MISSING!

```{r}


```
